cmake.md

```diff
@@ -0,0 +1,63 @@
+## CMake build system
+### Terminology
+
+The CMake system in symbiflow-arch-defs uses the following hierarchy:
+
+* `ARCH` - Architecture is the high level grouping of a set of "similar" FPGA
+  types. An example would be the "Lattice iCE40 Series" (ice40) or the "Xilinx Artix 7 Series" (artix7).
+* `DEVICE_TYPE` - A type of device.  This is defined as an `ARCH` and the
+  arch.xml which defines the units with this device type.
+* `DEVICE` - A specific instance of a `DEVICE_TYPE`.  A device definition is
+   a `DEVICE_TYPE`, a `PACKAGE` list, and IO map definitions for each package
+   if needed.  Each `DEVICE` should coorispond with a layout definition within
+   the `DEVICE_TYPE`.
+* `BOARD` - A specific `DEVICE` and `PACKAGE` instance, along with a
+  `PROG_TOOL`/`PROG_CMD` (e.g. commands to program a specified board).
+
+### Creating a new FPGA target
+
+Assuming that the board(s) you are interested in targetting is already
+defined, `ADD_FPGA_TARGET` is what will create targets.  See
+[`ADD_FPGA_TARGET` documentation](make/devices.cmake#L559).  `ADD_FPGA_TARGET`
+is roughly equivilant to CMake's `ADD_EXECUTABLE`.  Unlink `ADD_EXECUTABLE`,
+`ADD_FPGA_TARGET` requires that you specify which `BOARD` you are targetting.
+
+If you want to target multiple boards, `ADD_FPGA_TARGET_BOARDS` exists
+that will call `ADD_FPGA_TARGET` correctly for each board you request.  See
+[`ADD_FPGA_TARGET_BOARDS` documentation](make/devices.cmake#L458).
+
+#### Note on `ADD_FILE_TARGET` and `ADD_FPGA_BOARD`
+
+All source files in the symbiflow-arch-defs are required to have a file target
+associated with them.  This is done via
+[`ADD_FILE_TARGET`](make/file_targets.cmake#L193).  By default
+`ADD_FPGA_TARGET` and `ADD_FPGA_TARGET_BOARDS` will both implicitly invoke
+`ADD_FILE_TARGET` for you.  This handles the common case where input sources
+are not generated files.  However, if you want to generate files, then you
```

Comment:
Maybe?

`However, if you are using generated files, then you`

Reply:
Done